rpc/src/transport/channel.rs

```diff
@@ -113,33 +112,25 @@ mod tests {
             Ok::<_, io::Error>((response1, response2))
         };
 
-        let (response1, response2) = run_future(future::join(
+        let (response1, response2) = runtime::task::spawn(future::join(
             server,
             responses.unwrap_or_else(|e| panic!(e)),
         ))
+        .await
         .1;
 
         trace!("response1: {:?}, response2: {:?}", response1, response2);
 
-        assert!(response1.is_ok());
-        assert_eq!(response1.ok().unwrap(), 123);
+        let response1 = response1.map_err(|e| e.to_string());
+        ensure!(
```

Comment:
Let's use `assert_matches` instead? All of this ugly code I wrote before knowing about `assert_matches`.

Reply:
I am afraid that with assert failures will always output `the main future has panicked: Canceled`. Runtime makes us write tests that actually return an error - therefore we should use `ensure!` instead. In fact I am thinking of creating `ensure_eq!` and `ensure_matches!` and proposing them for inclusion into `failure` crate.

Reply:
I have a branch where I rewrote this test, and it seems to work as expected. In that branch, I spawn the server and make the client calls directly, not in an async block. I think the problem here is that runtime::spawn() returns a channel Receiver, which just sees a canceled sender.

Here's the code where I rewrote it: https://github.com/tikue/tarpc/blob/runtime/rpc/src/transport/channel.rs#L91-L118

Reply:
I took your version of the code and it spews the same nondescriptive `the main future has panicked: Canceled` when an assert fails.

Reply:
That's weird. Did you try checking out my branch? I changed Ok(123) to Ok(124) and got a reasonably nice assertion panic.

Reply:
I have adapted `assert_matches` to the standard propagation semantic.

https://github.com/vorot93/ensure_matches